### ISSUE TITLE
tf.sparse.softmax randomly outputs wrong results!

### ISSUE THREAD

**Issue by ArrowIntoTheSky** [tensorflow/tensorflow#53658]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Y
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.7.0
- Python version: 3.8
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source):N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Standalone code to reproduce the issue**
```
import tensorflow as tf
logits = tf.random.uniform([248, 248, 1, 16], dtype=tf.float32)
r1 = tf.nn.softmax(logits,axis=-1)
logits_sp = tf.sparse.from_dense(logits)
r2 = tf.sparse.softmax(logits_sp)
r3 = tf.sparse.to_dense(r2)
assert tf.math.reduce_all(tf.math.equal(r1, r3))
```
Normally the assertion would pass because `tf.sparse.softmax` and `tf.nn.softmax` should have the same output for the same input value. However, if I run the above code for 100 times then sometimes the assertion fails! Please see the [gist](https://colab.research.google.com/gist/ArrowIntoTheSky/fed8a0babf38a699ce40ecf59f70dbb6/untitled0.ipynb?authuser=2#scrollTo=PgfFQVFeCyLF) here for reference.

**Comment by sushreebarsa** [comment]
@Saduf2019 Was able to reproduce this issue on colab using TF version [2.7.0](https://colab.research.google.com/gist/sushreebarsa/69d2462eec2c137c257395327d2d96fa/untitled0.ipynb?authuser=2) and [tf-nightly](https://colab.research.google.com/gist/sushreebarsa/d531db2835f9e06d5a56e67eacab7995/untitled0.ipynb?authuser=2#scrollTo=3dMF6jtdC2zz)(2.9.0-dev20220108),please find the attached gists for reference.Thanks!

**Comment by cantonios** [comment]
The failures occur when a random "0" is encountered.  You can easily see different results with something like
```
logits = tf.constant([[
      0.8407991, 0., 0.15966213, 0.8176781
  ]], dtype=tf.float32)
  r1 = tf.nn.softmax(logits, axis=-1)
  
  logits_sp = tf.sparse.from_dense(logits)
  r2 = tf.sparse.softmax(logits_sp)
  r3 = tf.sparse.to_dense(r2)

  print(r1)
  print(r3)
```

Note that they do not produce the same thing _intentionally_.  For a dense softmax, zeros are considered, and the output of the softmax function will not be zero at those locations (since `exp(0) == 1`).  For sparse matrices, any zeroes [_do not participate_](https://www.tensorflow.org/api_docs/python/tf/sparse/softmax), and remain 0, removing them from both numerator and denominator.  This is to preserve sparsity (otherwise you would immediately have fully-dense matrices).
